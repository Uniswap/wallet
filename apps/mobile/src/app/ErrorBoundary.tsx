import React, { ErrorInfo } from 'react'
import { useTranslation } from 'react-i18next'
import RNRestart from 'react-native-restart'
import DeadLuni from 'src/assets/graphics/dead-luni.svg'
import { Button } from 'src/components/buttons/Button'
import { Flex } from 'src/components/layout'
import { Box } from 'src/components/layout/Box'
import { Text } from 'src/components/Text'
import { logger } from 'src/utils/logger'

interface ErrorBoundaryState {
  error: Error | null
}

// Uncaught errors during renders of subclasses will be caught here
// Errors in handlers (e.g. press handler) will not reach here
export class ErrorBoundary extends React.Component<unknown, ErrorBoundaryState> {
  constructor(props: unknown) {
    super(props)
    this.state = { error: null }
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    return { error }
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo): void {
    logger.error('ErrorBoundary', 'componentDidCatch', 'Error caught by boundary', error, errorInfo)
  }

  render(): React.ReactNode {
    const { error } = this.state
    if (error !== null) {
      return <ErrorScreen error={error} />
    }

    return this.props.children
  }
}

function ErrorScreen({ error }: { error: Error }): JSX.Element {
  const { t } = useTranslation()
  return (
    <Flex alignItems="center" flex={1} justifyContent="center" px="spacing16" py="spacing48">
      <Flex centered grow gap="spacing36">
        <DeadLuni />
        <Flex centered gap="spacing12">
          <Text variant="subheadLarge">{t('Uh oh!')}</Text>
          <Text variant="bodySmall">{t('Something crashed.')}</Text>
        </Flex>
        {error.message && __DEV__ && <Text variant="bodySmall">{error.message}</Text>}
      </Flex>
      <Box alignSelf="stretch">
        <Button
          label={t('Restart app')}
          onPress={(): void => {
            RNRestart.Restart()
          }}
        />
      </Box>
    </Flex>
  )
}
