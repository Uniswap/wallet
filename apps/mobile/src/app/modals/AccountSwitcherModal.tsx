import React, { useCallback, useMemo, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Alert } from 'react-native'
import 'react-native-gesture-handler'
import { Action } from 'redux'
import { useAppDispatch, useAppSelector, useAppTheme } from 'src/app/hooks'
import { navigate } from 'src/app/navigation/rootNavigation'
import { AccountList } from 'src/components/accounts/AccountList'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { Box, Flex } from 'src/components/layout'
import { Screen } from 'src/components/layout/Screen'
import { ActionSheetModal, MenuItemProp } from 'src/components/modals/ActionSheetModal'
import { BottomSheetModal } from 'src/components/modals/BottomSheetModal'
import { Text } from 'src/components/Text'
import { isICloudAvailable } from 'src/features/CloudBackup/RNICloudBackupsManager'
import { closeModal, openModal, selectModalState } from 'src/features/modals/modalSlice'
import { ImportType, OnboardingEntryPoint } from 'src/features/onboarding/utils'
import { ElementName, ModalName } from 'src/features/telemetry/constants'
import { OnboardingScreens, Screens } from 'src/screens/Screens'
import { openSettings } from 'src/utils/linking'
import PlusIcon from 'ui/src/assets/icons/plus.svg'
import { dimensions } from 'ui/src/theme/restyle/sizing'
import { createAccountActions } from 'wallet/src/features/wallet/create/createAccountSaga'
import {
  PendingAccountActions,
  pendingAccountActions,
} from 'wallet/src/features/wallet/create/pendingAccountsSaga'
import { useActiveAccountAddress, useNativeAccountExists } from 'wallet/src/features/wallet/hooks'
import { selectAllAccountsSorted } from 'wallet/src/features/wallet/selectors'
import { setAccountAsActive } from 'wallet/src/features/wallet/slice'

export function AccountSwitcherModal(): JSX.Element {
  const dispatch = useAppDispatch()
  const theme = useAppTheme()

  return (
    <BottomSheetModal
      disableSwipe
      backgroundColor={theme.colors.background1}
      name={ModalName.AccountSwitcher}
      onClose={(): Action => dispatch(closeModal({ name: ModalName.AccountSwitcher }))}>
      <Screen bg="background1" noInsets={true}>
        <AccountSwitcher
          onClose={(): void => {
            dispatch(closeModal({ name: ModalName.AccountSwitcher }))
          }}
        />
      </Screen>
    </BottomSheetModal>
  )
}

/**
 * Exported for testing only.
 * TODO [MOB-259] Once testing works with the BottomSheetModal stop exporting this component.
 */
export function AccountSwitcher({ onClose }: { onClose: () => void }): JSX.Element | null {
  const { t } = useTranslation()
  const theme = useAppTheme()

  const activeAccountAddress = useActiveAccountAddress()
  const dispatch = useAppDispatch()
  const hasImportedSeedPhrase = useNativeAccountExists()
  const modalState = useAppSelector(selectModalState(ModalName.AccountSwitcher))

  const [showAddWalletModal, setShowAddWalletModal] = useState(false)

  const accounts = useAppSelector(selectAllAccountsSorted)

  const onPressAccount = useCallback(
    (address: Address) => {
      onClose()
      // allow close modal logic to finish in JS thread before `setAccountAsActive` logic kicks in
      setImmediate(() => {
        dispatch(setAccountAsActive(address))
      })
    },
    [dispatch, onClose]
  )

  const onPressAddWallet = (): void => {
    setShowAddWalletModal(true)
  }

  const onCloseAddWallet = (): void => {
    setShowAddWalletModal(false)
  }

  const addWalletOptions = useMemo<MenuItemProp[]>(() => {
    const onPressCreateNewWallet = (): void => {
      // Clear any existing pending accounts first.
      dispatch(pendingAccountActions.trigger(PendingAccountActions.Delete))
      dispatch(createAccountActions.trigger())

      navigate(Screens.OnboardingStack, {
        screen: OnboardingScreens.EditName,
        params: {
          importType: hasImportedSeedPhrase ? ImportType.CreateAdditional : ImportType.CreateNew,
          entryPoint: OnboardingEntryPoint.Sidebar,
        },
      })
      setShowAddWalletModal(false)
      onClose()
    }

    const onPressAddViewOnlyWallet = (): void => {
      navigate(Screens.OnboardingStack, {
        screen: OnboardingScreens.WatchWallet,
        params: {
          importType: ImportType.Watch,
          entryPoint: OnboardingEntryPoint.Sidebar,
        },
      })
      setShowAddWalletModal(false)
      onClose()
    }

    const onPressImportWallet = (): void => {
      if (hasImportedSeedPhrase && activeAccountAddress) {
        dispatch(
          openModal({
            name: ModalName.RemoveWallet,
          })
        )
      } else {
        navigate(Screens.OnboardingStack, {
          screen: OnboardingScreens.SeedPhraseInput,
          params: { importType: ImportType.SeedPhrase, entryPoint: OnboardingEntryPoint.Sidebar },
        })
      }

      setShowAddWalletModal(false)
      onClose()
    }

    const onPressRestore = async (): Promise<void> => {
      const iCloudAvailable = await isICloudAvailable()

      if (!iCloudAvailable) {
        Alert.alert(
          t('iCloud Drive not available'),
          t(
            'Please verify that you are logged in to an Apple ID with iCloud Drive enabled on this device and try again.'
          ),
          [
            { text: t('Go to settings'), onPress: openSettings, style: 'default' },
            { text: t('Not now'), style: 'cancel' },
          ]
        )
        return
      }

      navigate(Screens.OnboardingStack, {
        screen: OnboardingScreens.RestoreCloudBackupLoading,
        params: { importType: ImportType.Restore, entryPoint: OnboardingEntryPoint.Sidebar },
      })
      setShowAddWalletModal(false)
      onClose()
    }

    const options = [
      {
        key: ElementName.CreateAccount,
        onPress: onPressCreateNewWallet,
        render: () => (
          <Box
            alignItems="center"
            borderBottomColor="backgroundOutline"
            borderBottomWidth={1}
            p="spacing16">
            <Text variant="bodyLarge">{t('Create a new wallet')}</Text>
          </Box>
        ),
      },
      {
        key: ElementName.AddViewOnlyWallet,
        onPress: onPressAddViewOnlyWallet,
        render: () => (
          <Box alignItems="center" p="spacing16">
            <Text variant="bodyLarge">{t('Add a view-only wallet')}</Text>
          </Box>
        ),
      },
      {
        key: ElementName.ImportAccount,
        onPress: onPressImportWallet,
        render: () => (
          <Box
            alignItems="center"
            borderTopColor="backgroundOutline"
            borderTopWidth={1}
            p="spacing16">
            <Text variant="bodyLarge">{t('Import a new wallet')}</Text>
          </Box>
        ),
      },
    ]

    if (!hasImportedSeedPhrase) {
      options.push({
        key: ElementName.RestoreFromICloud,
        onPress: onPressRestore,
        render: () => (
          <Box
            alignItems="center"
            borderTopColor="backgroundOutline"
            borderTopWidth={1}
            p="spacing16">
            <Text variant="bodyLarge">{t('Restore from iCloud')}</Text>
          </Box>
        ),
      })
    }

    return options
  }, [activeAccountAddress, dispatch, hasImportedSeedPhrase, onClose, t])

  if (!activeAccountAddress) {
    return null
  }

  const fullScreenContentHeight = 0.89 * dimensions.fullHeight

  return (
    <Flex fill gap="none" maxHeight={fullScreenContentHeight} mb="spacing12">
      <Flex row alignItems="center" borderBottomColor="backgroundOutline" mb="spacing16">
        <Box flex={1} pl="spacing24">
          <Text color="textPrimary" textAlign="left" variant="bodyLarge">
            {t('Your wallets')}
          </Text>
        </Box>
      </Flex>
      <AccountList accounts={accounts} isVisible={modalState.isOpen} onPress={onPressAccount} />
      <TouchableArea hapticFeedback my="spacing24" onPress={onPressAddWallet}>
        <Flex row alignItems="center" ml="spacing24">
          <Box
            borderColor="backgroundOutline"
            borderRadius="roundedFull"
            borderWidth={1}
            p="spacing12">
            <PlusIcon
              color={theme.colors.textPrimary}
              height={theme.iconSizes.icon16}
              strokeWidth={2}
              width={theme.iconSizes.icon16}
            />
          </Box>
          <Text color="textPrimary" variant="bodyLarge">
            {t('Add wallet')}
          </Text>
        </Flex>
      </TouchableArea>

      <ActionSheetModal
        isVisible={showAddWalletModal}
        name={ModalName.AddWallet}
        options={addWalletOptions}
        onClose={onCloseAddWallet}
      />
    </Flex>
  )
}
