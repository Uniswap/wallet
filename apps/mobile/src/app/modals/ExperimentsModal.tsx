import React from 'react'
import { Action } from 'redux'
import { useAppDispatch } from 'src/app/hooks'
import { Switch } from 'src/components/buttons/Switch'
import { Flex } from 'src/components/layout/Flex'
import { BottomSheetModal } from 'src/components/modals/BottomSheetModal'
import { Text } from 'src/components/Text'
import { FEATURE_FLAGS } from 'src/features/experiments/constants'
import { useFeatureFlag } from 'src/features/experiments/hooks'
import { closeModal } from 'src/features/modals/modalSlice'
import { ModalName } from 'src/features/telemetry/constants'
import { Statsig } from 'statsig-react'

export function ExperimentsModal(): JSX.Element {
  const dispatch = useAppDispatch()

  return (
    <BottomSheetModal
      name={ModalName.Experiments}
      onClose={(): Action => dispatch(closeModal({ name: ModalName.Experiments }))}>
      <Flex
        gap="spacing16"
        justifyContent="flex-start"
        pb="spacing36"
        pt="spacing12"
        px="spacing24">
        <Flex gap="spacing8">
          <Text variant="subheadLarge">⛳️ Feature Flags</Text>
          <Text variant="bodySmall">
            Overridden feature flags are reset when the app is restarted
          </Text>
        </Flex>
        {Object.values(FEATURE_FLAGS).map((featureFlag) => {
          return <FeatureFlagRow key={featureFlag} featureFlag={featureFlag} />
        })}
      </Flex>
    </BottomSheetModal>
  )
}

function FeatureFlagRow({ featureFlag }: { featureFlag: FEATURE_FLAGS }): JSX.Element {
  const status = useFeatureFlag(featureFlag)

  return (
    <Flex row alignItems="center" justifyContent="space-between">
      <Text variant="bodyLarge">{featureFlag}</Text>
      <Switch
        value={status}
        onValueChange={(newValue: boolean): void => {
          Statsig.overrideGate(featureFlag, newValue)
        }}
      />
    </Flex>
  )
}
