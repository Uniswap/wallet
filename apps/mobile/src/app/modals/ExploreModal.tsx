import React from 'react'
import { useAppDispatch, useAppTheme } from 'src/app/hooks'
import { ExploreStackNavigator } from 'src/app/navigation/navigation'
import { ExploreStackParamList } from 'src/app/navigation/types'
import { BottomSheetModal } from 'src/components/modals/BottomSheetModal'
import { closeModal } from 'src/features/modals/modalSlice'
import { ModalName } from 'src/features/telemetry/constants'
import { Screens } from 'src/screens/Screens'

type InnerExploreStackParamList = Omit<ExploreStackParamList, Screens.Explore>

// The ExploreModalState allows a Screen and its Params to be defined, except for the initial Explore screen.
// This workaround facilitates navigation to any screen within the ExploreStack from outside.
// Implementation of this lives inside screens/ExploreScreen
export type ExploreModalState = {
  [V in keyof InnerExploreStackParamList]: { screen: V; params: InnerExploreStackParamList[V] }
}[keyof InnerExploreStackParamList]

export function ExploreModal(): JSX.Element {
  const theme = useAppTheme()
  const appDispatch = useAppDispatch()

  const onClose = (): void => {
    appDispatch(closeModal({ name: ModalName.Explore }))
  }

  return (
    <BottomSheetModal
      blurredBackground
      fullScreen
      hideKeyboardOnDismiss
      backgroundColor={theme.colors.none}
      hideHandlebar={true}
      name={ModalName.Explore}
      renderBehindInset={true}
      onClose={onClose}>
      <ExploreStackNavigator />
    </BottomSheetModal>
  )
}
