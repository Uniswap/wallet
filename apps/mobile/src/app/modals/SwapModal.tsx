import React, { useCallback } from 'react'
import { useAppDispatch, useAppSelector, useAppTheme } from 'src/app/hooks'
import { BottomSheetModal } from 'src/components/modals/BottomSheetModal'
import { closeModal, selectModalState } from 'src/features/modals/modalSlice'
import { ModalName } from 'src/features/telemetry/constants'
import { SwapFlow } from 'src/features/transactions/swap/SwapFlow'

export function SwapModal(): JSX.Element {
  const theme = useAppTheme()
  const appDispatch = useAppDispatch()
  const modalState = useAppSelector(selectModalState(ModalName.Swap))

  const onClose = useCallback((): void => {
    appDispatch(closeModal({ name: ModalName.Swap }))
  }, [appDispatch])

  return (
    <BottomSheetModal
      fullScreen
      hideKeyboardOnDismiss
      backgroundColor={theme.colors.background1}
      name={ModalName.Swap}
      onClose={onClose}>
      <SwapFlow prefilledState={modalState.initialState} onClose={onClose} />
    </BottomSheetModal>
  )
}
