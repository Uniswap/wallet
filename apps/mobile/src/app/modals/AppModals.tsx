import React from 'react'
import { AccountSwitcherModal } from 'src/app/modals/AccountSwitcherModal'
import { ExperimentsModal } from 'src/app/modals/ExperimentsModal'
import { ExploreModal } from 'src/app/modals/ExploreModal'
import { SwapModal } from 'src/app/modals/SwapModal'
import { TransferTokenModal } from 'src/app/modals/TransferTokenModal'
import { LazyModalRenderer } from 'src/app/modals/utils'
import { ForceUpgradeModal } from 'src/components/forceUpgrade/ForceUpgradeModal'
import { RemoveWalletModal } from 'src/components/RemoveWallet/RemoveWalletModal'
import { RestoreWalletModal } from 'src/components/RestoreWalletModal/RestoreWalletModal'
import { WalletConnectModals } from 'src/components/WalletConnect/WalletConnectModals'
import { LockScreenModal } from 'src/features/authentication/LockScreenModal'
import { FiatOnRampModal } from 'src/features/fiatOnRamp/FiatOnRampModal'
import { ModalName } from 'src/features/telemetry/constants'
import { SettingsFiatCurrencyModal } from 'src/screens/SettingsFiatCurrencyModal'
import { SettingsLanguageModal } from 'src/screens/SettingsLanguageModal'

export function AppModals(): JSX.Element {
  return (
    <>
      <LazyModalRenderer name={ModalName.Experiments}>
        <ExperimentsModal />
      </LazyModalRenderer>

      <LazyModalRenderer name={ModalName.FiatOnRamp}>
        <FiatOnRampModal />
      </LazyModalRenderer>

      <LazyModalRenderer name={ModalName.Explore}>
        <ExploreModal />
      </LazyModalRenderer>

      <ForceUpgradeModal />

      <LockScreenModal />

      <LazyModalRenderer name={ModalName.Swap}>
        <SwapModal />
      </LazyModalRenderer>

      <LazyModalRenderer name={ModalName.Send}>
        <TransferTokenModal />
      </LazyModalRenderer>

      <WalletConnectModals />

      <LazyModalRenderer name={ModalName.AccountSwitcher}>
        <AccountSwitcherModal />
      </LazyModalRenderer>

      <LazyModalRenderer name={ModalName.RemoveWallet}>
        <RemoveWalletModal />
      </LazyModalRenderer>

      <LazyModalRenderer name={ModalName.RestoreWallet}>
        <RestoreWalletModal />
      </LazyModalRenderer>

      <LazyModalRenderer name={ModalName.LanguageSelector}>
        <SettingsLanguageModal />
      </LazyModalRenderer>

      <LazyModalRenderer name={ModalName.FiatCurrencySelector}>
        <SettingsFiatCurrencyModal />
      </LazyModalRenderer>
    </>
  )
}
