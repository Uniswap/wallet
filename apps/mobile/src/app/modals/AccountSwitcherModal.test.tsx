import { MockedResponse } from '@apollo/client/testing'
import React from 'react'
import { PreloadedState } from 'redux'
import { AccountSwitcher } from 'src/app/modals/AccountSwitcherModal'
import { RootState } from 'src/app/rootReducer'
import { AccountListDocument, AccountListQuery } from 'src/data/__generated__/types-and-hooks'
import { initialModalState } from 'src/features/modals/modalSlice'
import { ModalName } from 'src/features/telemetry/constants'
import { ACCOUNT_ADDRESS_ONE, mockWalletPreloadedState } from 'src/test/fixtures'
import { Portfolios } from 'src/test/gqlFixtures'
import { render } from 'src/test/test-utils'

jest.useFakeTimers()

const preloadedState = {
  ...mockWalletPreloadedState,
  modals: {
    ...initialModalState,
    [ModalName.AccountSwitcher]: { isOpen: true },
  },
} as unknown as PreloadedState<RootState>

const AccountListMock: MockedResponse<AccountListQuery> = {
  request: {
    query: AccountListDocument,
    variables: {
      addresses: [ACCOUNT_ADDRESS_ONE],
    },
  },
  result: {
    data: {
      portfolios: Portfolios,
    },
  },
}

// TODO [MOB-3961]: Figure out how to do snapshot tests when there is a BottomSheetModal
describe(AccountSwitcher, () => {
  it('renders correctly', () => {
    const tree = render(
      <AccountSwitcher
        onClose={(): void => {
          return
        }}
      />,
      { preloadedState, mocks: [AccountListMock] }
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
