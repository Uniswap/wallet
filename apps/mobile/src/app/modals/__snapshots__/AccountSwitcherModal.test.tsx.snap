// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountSwitcher renders correctly 1`] = `
<View
  sentry-label="Flex"
  style={
    [
      {
        "alignItems": "stretch",
        "display": "flex",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "flex-start",
        "marginBottom": 12,
        "maxHeight": 1187.26,
      },
    ]
  }
>
  <View
    sentry-label="Flex"
    style={
      [
        {
          "alignItems": "center",
          "borderBottomColor": "#D2D9EE",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginBottom": 16,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "display": "flex",
            "flex": 1,
            "paddingLeft": 24,
          },
        ]
      }
    >
      <Text
        allowFontScaling={true}
        maxFontSizeMultiplier={1.4}
        style={
          [
            {
              "color": "#0E1524",
              "display": "flex",
              "fontFamily": "Inter-Medium",
              "fontSize": 17,
              "lineHeight": 24,
              "textAlign": "left",
            },
          ]
        }
      >
        Your wallets
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "display": "flex",
          "flexGrow": 0,
          "flexShrink": 0,
          "height": 0,
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "display": "flex",
          "flexShrink": 1,
          "position": "relative",
        },
      ]
    }
  >
    ExpoLinearGradient
    <RCTScrollView
      bounces={false}
      collapsable={false}
      handlerTag={1}
      handlerType="NativeViewGestureHandler"
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      scrollEnabled={false}
      showsVerticalScrollIndicator={false}
      waitFor={
        [
          {
            "current": null,
          },
        ]
      }
    >
      <View>
        <View
          actions={
            [
              {
                "systemIcon": "doc.on.doc",
                "title": "Copy wallet address",
              },
              {
                "systemIcon": "gearshape",
                "title": "Wallet settings",
              },
              {
                "destructive": true,
                "systemIcon": "trash",
                "title": "Remove wallet",
              },
            ]
          }
          onPress={[Function]}
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 5,
                "left": 5,
                "right": 5,
                "top": 5,
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "display": "flex",
                "opacity": 1,
                "paddingBottom": 12,
                "paddingHorizontal": 24,
                "paddingTop": 8,
              }
            }
          >
            <View
              sentry-label="Flex"
              style={
                [
                  {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  },
                ]
              }
              testID="account_item/0x82D56A352367453f74FC0dC7B071b311da373Fa6"
            >
              <View
                sentry-label="Flex"
                style={
                  [
                    {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                      "flexShrink": 1,
                      "justifyContent": "flex-start",
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "display": "flex",
                        "height": "100%",
                        "position": "relative",
                      },
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {
                          "backgroundColor": "transparent",
                          "borderColor": "transparent",
                          "borderRadius": 999999,
                          "borderWidth": 0,
                          "display": "flex",
                          "position": "relative",
                        },
                        {
                          "padding": 0,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          {
                            "display": "flex",
                            "height": 40,
                            "width": 40,
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          [
                            {
                              "display": "flex",
                              "height": 40,
                              "width": 40,
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            {
                              "flex": 1,
                            }
                          }
                        >
                          <View
                            clip={true}
                            mask={
                              <UniconMask
                                attributeData={
                                  {
                                    "0": "#6100FF",
                                    "1": "#76D191",
                                    "2": [
                                      {
                                        "fillType": "evenOdd",
                                        "path": "M31 33V16C31 8.8203 25.1797 3 18 3C10.8203 3 5 8.8203 5 16V33H31ZM18 0C9.16344 0 2 7.16344 2 16V35.807C2 35.9136 2.08642 36 2.19303 36H33.807C33.9136 36 34 35.9136 34 35.807V16C34 7.16344 26.8366 0 18 0Z",
                                      },
                                    ],
                                    "3": [
                                      {
                                        "path": "M8 8C8 8 13.1678 6.70089 14.2197 5.649C15.2716 4.59712 15.2583 2.87836 14.19 1.81004C13.1216 0.741719 11.4029 0.728395 10.351 1.78028C9.29911 2.83216 8 8 8 8Z",
                                      },
                                      {
                                        "path": "M8 8C8 8 13.1678 9.29911 14.2197 10.351C15.2716 11.4029 15.2583 13.1216 14.19 14.19C13.1216 15.2583 11.4029 15.2716 10.351 14.2197C9.29911 13.1678 8 8 8 8Z",
                                      },
                                      {
                                        "path": "M1.78028 10.351C2.83216 9.29911 8 8 8 8C8 8 6.70089 13.1678 5.649 14.2197C4.59712 15.2716 2.87836 15.2583 1.81004 14.19C0.741719 13.1216 0.728395 11.4029 1.78028 10.351Z",
                                      },
                                      {
                                        "path": "M8 8C8 8 2.83216 6.70089 1.78028 5.649C0.728395 4.59712 0.741719 2.87836 1.81004 1.81004C2.87836 0.741719 4.59712 0.728395 5.649 1.78028C6.70089 2.83216 8 8 8 8Z",
                                      },
                                    ],
                                  }
                                }
                                showBorder={false}
                                size={40}
                              />
                            }
                          >
                            <View
                              transform={
                                [
                                  {
                                    "scale": 1.0810810810810811,
                                  },
                                ]
                              }
                            >
                              <View
                                height={37}
                                width={37}
                                x={0}
                                y={0}
                              >
                                <View
                                  colors={
                                    [
                                      "#6100FF",
                                      "#76D191",
                                    ]
                                  }
                                  end={null}
                                  start={null}
                                />
                              </View>
                              <View
                                color="#D3EBA3"
                                cx={18.5}
                                cy={-13.36111111111111}
                                r={30.833333333333336}
                              >
                                <View
                                  blur={15}
                                />
                              </View>
                            </View>
                          </View>
                          <View
                            clip={true}
                            mask={
                              <UniconMask
                                attributeData={
                                  {
                                    "0": "#6100FF",
                                    "1": "#76D191",
                                    "2": [
                                      {
                                        "fillType": "evenOdd",
                                        "path": "M31 33V16C31 8.8203 25.1797 3 18 3C10.8203 3 5 8.8203 5 16V33H31ZM18 0C9.16344 0 2 7.16344 2 16V35.807C2 35.9136 2.08642 36 2.19303 36H33.807C33.9136 36 34 35.9136 34 35.807V16C34 7.16344 26.8366 0 18 0Z",
                                      },
                                    ],
                                    "3": [
                                      {
                                        "path": "M8 8C8 8 13.1678 6.70089 14.2197 5.649C15.2716 4.59712 15.2583 2.87836 14.19 1.81004C13.1216 0.741719 11.4029 0.728395 10.351 1.78028C9.29911 2.83216 8 8 8 8Z",
                                      },
                                      {
                                        "path": "M8 8C8 8 13.1678 9.29911 14.2197 10.351C15.2716 11.4029 15.2583 13.1216 14.19 14.19C13.1216 15.2583 11.4029 15.2716 10.351 14.2197C9.29911 13.1678 8 8 8 8Z",
                                      },
                                      {
                                        "path": "M1.78028 10.351C2.83216 9.29911 8 8 8 8C8 8 6.70089 13.1678 5.649 14.2197C4.59712 15.2716 2.87836 15.2583 1.81004 14.19C0.741719 13.1216 0.728395 11.4029 1.78028 10.351Z",
                                      },
                                      {
                                        "path": "M8 8C8 8 2.83216 6.70089 1.78028 5.649C0.728395 4.59712 0.741719 2.87836 1.81004 1.81004C2.87836 0.741719 4.59712 0.728395 5.649 1.78028C6.70089 2.83216 8 8 8 8Z",
                                      },
                                    ],
                                  }
                                }
                                overlay={true}
                                showBorder={false}
                                size={40}
                              />
                            }
                          >
                            <View
                              color="#000000"
                              height={40}
                              opacity={0.08}
                              width={40}
                              x={0}
                              y={0}
                            />
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
                <View
                  style={
                    [
                      {
                        "display": "flex",
                        "flexGrow": 0,
                        "flexShrink": 0,
                        "height": 16,
                        "width": 16,
                      },
                    ]
                  }
                />
                <View
                  sentry-label="Flex"
                  style={
                    [
                      {
                        "alignItems": "stretch",
                        "display": "flex",
                        "flex": 1,
                        "flexDirection": "column",
                        "justifyContent": "flex-start",
                      },
                    ]
                  }
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      [
                        {
                          "color": "#0E1524",
                          "display": "flex",
                          "fontFamily": "Inter-Medium",
                          "fontSize": 20,
                          "lineHeight": 24,
                        },
                      ]
                    }
                  >
                    Test Account
                  </Text>
                  <View
                    style={
                      [
                        {
                          "display": "flex",
                          "flexGrow": 0,
                          "flexShrink": 0,
                          "height": 0,
                        },
                      ]
                    }
                  />
                  <View
                    onLayout={[Function]}
                  >
                    <View
                      style={
                        [
                          {
                            "alignItems": "center",
                            "display": "flex",
                            "flexDirection": "row",
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          [
                            {
                              "alignItems": "center",
                              "display": "flex",
                              "flexDirection": "row",
                              "position": "relative",
                            },
                          ]
                        }
                      >
                        <View
                          accessibilityElementsHidden={true}
                          importantForAccessibility="no-hide-descendants"
                        >
                          <Text
                            allowFontScaling={true}
                            maxFontSizeMultiplier={1.4}
                            style={
                              [
                                {
                                  "color": "#888FAB",
                                  "display": "flex",
                                  "fontFamily": "Inter-Medium",
                                  "fontSize": 15,
                                  "lineHeight": 20,
                                  "opacity": 0,
                                },
                              ]
                            }
                          >
                            $000.00
                          </Text>
                        </View>
                        <View
                          style={
                            [
                              {
                                "backgroundColor": "#E8ECFB",
                                "borderRadius": 4,
                                "bottom": "5%",
                                "display": "flex",
                                "left": 0,
                                "position": "absolute",
                                "right": 0,
                                "top": "5%",
                              },
                            ]
                          }
                        />
                      </View>
                    </View>
                  </View>
                </View>
              </View>
              <View
                style={
                  [
                    {
                      "display": "flex",
                      "flexGrow": 0,
                      "flexShrink": 0,
                      "height": 16,
                      "width": 16,
                    },
                  ]
                }
              />
              <View
                sentry-label="Flex"
                style={
                  [
                    {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                    },
                  ]
                }
              >
                <svg
                  color="#FC72FF"
                  data-file-name="SvgCheck"
                  height={24}
                  width={24}
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    ExpoLinearGradient
  </View>
  <View
    style={
      [
        {
          "display": "flex",
          "flexGrow": 0,
          "flexShrink": 0,
          "height": 0,
        },
      ]
    }
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={
      {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "display": "flex",
        "marginVertical": 24,
        "opacity": 1,
      }
    }
  >
    <View
      sentry-label="Flex"
      style={
        [
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginLeft": 24,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "borderColor": "#D2D9EE",
              "borderRadius": 999999,
              "borderWidth": 1,
              "display": "flex",
              "padding": 12,
            },
          ]
        }
      >
        <svg
          color="#0E1524"
          data-file-name="SvgPlus"
          height={16}
          strokeWidth={2}
          width={16}
        />
      </View>
      <View
        style={
          [
            {
              "display": "flex",
              "flexGrow": 0,
              "flexShrink": 0,
              "height": 16,
              "width": 16,
            },
          ]
        }
      />
      <Text
        allowFontScaling={true}
        maxFontSizeMultiplier={1.4}
        style={
          [
            {
              "color": "#0E1524",
              "display": "flex",
              "fontFamily": "Inter-Medium",
              "fontSize": 17,
              "lineHeight": 24,
            },
          ]
        }
      >
        Add wallet
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "display": "flex",
          "flexGrow": 0,
          "flexShrink": 0,
          "height": 0,
        },
      ]
    }
  />
</View>
`;
